### PR TITLE
JENKINS-48392 record the name of the plugin used

### PR DESCRIPTION
**Pull request by i386**
# Description

* Record the name of the plugin used
* Use our own chunk visitor that has less overhead than the one Blue Ocean uses

See [JENKINS-48392](https://issues.jenkins-ci.org/browse/JENKINS-48392).

# Submitter checklist
- [ ] Link to JIRA ticket in description, if appropriate.
- [ ] Change is code complete and matches issue description
- [ ] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

### CODE REVIEW

File: blueocean-pipeline-api-impl/src/test/java/io/jenkins/blueocean/rest/impl/pipeline/analytics/PipelinePluginAnalyticsTest.java [hunk 1]
```diff
@@ -33,6 +33,7 @@ public void testGenerateAnalyticsEventWithDeclarative() throws Exception {
         Assert.assertEquals("timesUsed", 2, properties.get("timesUsed"));
         Assert.assertEquals("isDeclarative", true, properties.get("isDeclarative"));
         Assert.assertEquals("runResult", "SUCCESS", properties.get("runResult"));
+//        Assert.assertEquals("plugin", "", properties.get("plugin"));
```
Comment (i386) [inline blueocean-pipeline-api-impl/src/test/java/io/jenkins/blueocean/rest/impl/pipeline/analytics/PipelinePluginAnalyticsTest.java]: so it turns out `pluginManager.whichPlugin` doesn't work inside of JenkinsRule tests but does work when you run the app normally. Go figure :(

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/analytics/PipelinePluginAnalytics.java [hunk 2]
```diff
@@ -27,23 +43,51 @@ public void onCompleted(WorkflowRun workflowRun, @Nonnull TaskListener listener)
...
             props.put("timesUsed", value);
             props.put("isDeclarative", isDeclarative);
             props.put("runResult", resultAsString);
+            String pluginName = visitor.stepToPlugin.get(key);
+            props.put("plugin", pluginName != null ? pluginName : "unknown");
             analytics.track(new TrackRequest("pipeline_step_used", props.build()));
         });
     }
+
+    private class PipelineStepVisitor extends StandardChunkVisitor {
```
Comment (i386) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/analytics/PipelinePluginAnalytics.java]: @abayer decided it was more efficient to do things with my own chunk visitor

File: blueocean-pipeline-api-impl/src/test/java/io/jenkins/blueocean/rest/impl/pipeline/analytics/PipelinePluginAnalyticsTest.java (hunk 1 above)
Comment (i386) [reply blueocean-pipeline-api-impl/src/test/java/io/jenkins/blueocean/rest/impl/pipeline/analytics/PipelinePluginAnalyticsTest.java]: Any ideas @vivek or @jglick ?

**Comment by i386** [comment]
Thanks @abayer

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/analytics/PipelinePluginAnalytics.java [hunk 3]
```diff
@@ -27,23 +43,51 @@ public void onCompleted(WorkflowRun workflowRun, @Nonnull TaskListener listener)
...
             props.put("timesUsed", value);
             props.put("isDeclarative", isDeclarative);
             props.put("runResult", resultAsString);
+            String pluginName = visitor.stepToPlugin.get(key);
+            props.put("plugin", pluginName != null ? pluginName : "unknown");
             analytics.track(new TrackRequest("pipeline_step_used", props.build()));
         });
     }
+
+    private static final class PipelineStepVisitor extends StandardChunkVisitor {
```
Comment (svanoort) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/analytics/PipelinePluginAnalytics.java]: If you're just collecting a Tally, use https://github.com/jenkinsci/workflow-api-plugin/blob/master/src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/FlowNodeVisitor.java and 
https://github.com/jenkinsci/workflow-api-plugin/blob/master/src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/AbstractFlowScanner.java#L373

Bismuth is meant for this kind of work, and meant to make it simple -- you need not just collect on AtomNodes, we can get info on which block-scoped steps were run as well -- caveat: you might get some duplication due to nesting of blocks.

I'd actually imagined doing counts of steps by type but nobody asked for it, so I didn't code it up.

Comment (svanoort) [reply blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/analytics/PipelinePluginAnalytics.java]: On second thought, better approach here is probably to use a GraphListener so you're not doing a complete extra scan through the FlowGraph.

File: blueocean-pipeline-api-impl/src/test/java/io/jenkins/blueocean/rest/impl/pipeline/analytics/PipelinePluginAnalyticsTest.java (hunk 1 above)
Comment (svanoort) [reply blueocean-pipeline-api-impl/src/test/java/io/jenkins/blueocean/rest/impl/pipeline/analytics/PipelinePluginAnalyticsTest.java]: @Jesse claimed to add something to the ArgumentsActionImpl that will give you a more direct way to get at this and may not need the plugin manager -- I haven't tried it out though.

**Comment by svanoort** [comment]
Left a couple comments -- in general I think this approach *should* give a result but I think the way it's structured could use a bit of polish.

**Comment by vivek** [comment]
@i386 Analytics has been disabled in blueocean: https://github.com/jenkinsci/blueocean-plugin/pull/1618. Is this PR relevant any more?

**Comment by i386** [comment]
@vivek still relevant. I haven't had the chance to circle back yet.

**Comment by vivek** [comment]
@i386 ok, thanks.

**Pull request closed without merging**